services/sqs/src/main/java/software/amazon/awssdk/services/sqs/internal/batchmanager/AsyncReceiveMessageBatch.java

```diff
@@ -0,0 +1,173 @@
+                                  }
+                                  open.set(true);
+                                  return this;
+                              });
+        } finally {
+            visibilityDeadlineNano = System.nanoTime() + visibilityTimeout.toNanos();
+        }
+
+    }
+
+    public boolean isEmpty() {
+        return messages == null || messages.isEmpty();
+    }
+
+    public Throwable getException() {
+        checkIfOpen();
+        return exception;
+    }
+
+    public void setException(Throwable exception) {
+        this.exception = exception;
+    }
+
+    public Message removeMessage() {
+        checkIfOpen();
+        if (isExpired()) {
+            clear();
+            return null;
+        }
+        return messages.isEmpty() ? null : messages.remove(messages.size() - 1);
+    }
+
+    public boolean isExpired() {
+        return System.nanoTime() > visibilityDeadlineNano;
+    }
+
+    public void clear() {
+        if (!isEmpty()) {
+            Optional.ofNullable(nackMessages())
+                    .ifPresent(future -> future.whenComplete((r, t) -> {
+                        // Logging an error is sufficient here as this is an asynchronous cleanup activity,
+                        // and there are no dependent tasks waiting for its completion.
+                        if (t != null) {
+                            log.error(() -> "Could not change visibility for queue " + queueUrl, t);
+                        }
+                    }));
+            messages.clear();
+        }
+    }
+
+    private CompletableFuture<ChangeMessageVisibilityBatchResponse> nackMessages() {
+        if (messages == null || messages.isEmpty()) {
+            return null;
+        }
+
+        List<ChangeMessageVisibilityBatchRequestEntry> entries =
+            IntStream.range(0, messages.size())
+                     .mapToObj(i -> ChangeMessageVisibilityBatchRequestEntry.builder()
+                                                                            .id(String.valueOf(i))
+                                                                            .receiptHandle(messages.get(i).receiptHandle())
+                                                                            .visibilityTimeout(0)
+                                                                            .build())
+                     .collect(Collectors.toList());
+
+        ChangeMessageVisibilityBatchRequest batchRequest = ChangeMessageVisibilityBatchRequest.builder()
+                                                                                              .queueUrl(queueUrl)
+                                                                                              .entries(entries)
+                                                                                              .build();
+
+        return asyncClient.changeMessageVisibilityBatch(batchRequest);
+    }
+
+    private void checkIfOpen() {
```

Comment:
I'm not really sure what the purpose of ensure this is "open" should be. What does "open" indicate?